proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,262 @@
+<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->
+
+# Improvements to Kroxylicious (Proxy) Metrics
+
+The document aims to:
+
+* describe the short-comings with the current metrics
+* identify use-cases for Kroxylicious metrics
+* propose improvements.
+* propose an implementation plan.
+
+<!-- TOC -->
+* [Improvements to Kroxylicious (Proxy) Metrics](#improvements-to-kroxylicious-proxy-metrics)
+  * [Current situation](#current-situation)
+    * [Specific short-comings/weaknesses](#specific-short-comingsweaknesses)
+    * [General short-comings/weaknesses](#general-short-comingsweaknesses)
+  * [Use Cases](#use-cases)
+  * [Proposal](#proposal)
+  * [Summary](#summary)
+    * [New counter `kroxylicious_downstream_kafka_message_total`](#new-counter-kroxyliciousdownstreamkafkamessagetotal)
+    * [New counter `kroxylicious_upstream_kafka_message_total`](#new-counter-kroxyliciousupstreamkafkamessagetotal)
+    * [New distribution `kroxylicious_downstream_kafka_message_size_bytes`](#new-distribution-kroxyliciousdownstreamkafkamessagesizebytes)
+    * [New distribution `kroxylicious_upstream_kafka_message_size_bytes`](#new-distribution-kroxyliciousupstreamkafkamessagesizebytes)
+    * [New distribution `kroxylicious_message_process_time`](#new-distribution-kroxyliciousmessageprocesstime)
+    * [New gauge metrics `kroxylicious_(up|down)stream_backpressure_state`](#new-gauge-metrics-kroxyliciousupdownstreambackpressurestate)
+    * [New distribution `kroxylicious_(up|down)stream_backpressure_time`](#new-distribution-kroxyliciousupdownstreambackpressuretime)
+    * [Add label `node_id` and `virtual_cluster` to `kroxylicious_(down|up)stream_connections_attempts` and `kroxylicious_(down|up)stream_errors`](#add-label-nodeid-and-virtualcluster-to-kroxyliciousdownupstreamconnectionsattempts-and-kroxyliciousdownupstreamerrors)
```

Comment:
Are we going to have the `nodeID` only for the `kroxylicious_(down|up)stream_connections_attempts` and `kroxylicious_(down|up)stream_errors` ? Reading the proposal further I think we are going to add these labels to other metrics also. Reading the sentence it felt like that we are adding the labels to only these specific metrics

Reply:
> Are we going to have the nodeID only for the kroxylicious_(down|up)stream_connections_attempts and kroxylicious_(down|up)stream_errors ?

no.  `node_id` will be present on all the new metrics where it is appropriate (that's every one except the backpressure metric).

>  Reading the proposal further I think we are going to add these labels to other metrics also. 

Yup.

> Reading the sentence it felt like that we are adding the labels to only these specific metrics

no, I spilt up the summary thus:

- new metrics
- changes to existing metrics that we are keeping
- metrics we are deprecating

If that's not clear I can improve.